### PR TITLE
Fix machine deletion

### PR DESCRIPTION
**Pull request by vikaschoudhary16**
**What this PR does / why we need it**:
Currently if actuator fails to create a machine and user wants to delete the machine object, kubectl hangs forever. This is because in the reconcile function, error is thrown because of failing to find machine. Kubectl hangs because of the `finalizer`.

This PR adds logic to first verify if machine exists at the actuator before executing Delete at  actuator. 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**Release note**:

```release-note

```
cc @roberthbailey @pwittrock

### CODE REVIEW

File: pkg/controller/machine/controller.go [hunk 1]
```diff
@@ -97,7 +97,13 @@ type ReconcileMachine struct {
 func (r *ReconcileMachine) Reconcile(request reconcile.Request) (reconcile.Result, error) {
 	// Fetch the Machine instance
 	m := &clusterv1.Machine{}
-	err := r.Get(context.Background(), request.NamespacedName, m)
+	cluster, err := r.getCluster(m)
+	if err != nil {
+		glog.Errorf("Error in getting cluster: %+v", err)
```
Comment (frobware) [inline pkg/controller/machine/controller.go]: Any reason to use `%+v` over `%v`?

File: pkg/controller/machine/controller.go [hunk 2]
```diff
@@ -134,11 +140,18 @@ func (r *ReconcileMachine) Reconcile(request reconcile.Request) (reconcile.Resul
 			return reconcile.Result{}, nil
 		}
 		glog.Infof("reconciling machine object %v triggers delete.", name)
-		if err := r.delete(m); err != nil {
-			glog.Errorf("Error deleting machine object %v; %v", name, err)
+
+		exist, err := r.actuator.Exists(cluster, m)
+		if err != nil {
+			glog.Errorf("Failed to delete. Error checking existence of machine instance for machine object %v; %v", name, err)
```
Comment (frobware) [inline pkg/controller/machine/controller.go]: nit: use `%v: %v` for formatting errors.

File: pkg/controller/machine/controller.go [hunk 3]
```diff
@@ -134,11 +140,18 @@ func (r *ReconcileMachine) Reconcile(request reconcile.Request) (reconcile.Resul
...
 			return reconcile.Result{}, err
 		}
-
+		if exist {
+			if err := r.delete(m); err != nil {
+				glog.Errorf("Error deleting machine object %v; %v", name, err)
```
Comment (frobware) [inline pkg/controller/machine/controller.go]: nit: use `%v: %v` for formatting errors.

**Comment by davidewatson** [comment]
/lgtm cancel

I think I was mistaken before. Since the actuator implements both Delete and Exists maybe it should be the responsibility of the provider to succeed when deleting a machine which _it_ does not think exists.

Which provider do you see this behavor with?

**Comment by vikaschoudhary16** [comment]
>I think I was mistaken before. Since the actuator implements both Delete and Exists maybe it should be > the responsibility of the provider to succeed when deleting a machine which it does not think exists.

sounds reasonable. Let me check.

> Which provider do you see this behavor with?

libvirt

**Comment by vikaschoudhary16** [comment]
It makes more sense to fix this at actuator. So closing this.

**Pull request closed without merging**